### PR TITLE
Add a flag to tests to start cchost.

### LINKED ISSUE #6: Remove dependency on `sandbox.sh`
CCF ships a script `sandbox.sh` to easily launch CCF apps for testing purposes. We currently use this within `start.sh` for manual testing and indirectly for the functional tests (by running `start.sh`), both when not using Docker. For Docker, we have `docker/run-dev.sh` which does roughly the same thing without relying on `sandbox.sh`.

There are a couple of disadvantages using `sandbox.sh`:
- It starts the network with a default constitution that doesn't require voting on proposals. This makes testing of client code dealing with governance hard.
- It does a bunch of magic to get the network running. Again, this makes testing of those parts hard.
- It's not usable with our Docker images. This means there will be two ways to run tests that will differ more than necessary.
- It creates its own Python venv which takes a while to get ready as it's installing dependencies. This means using it from tests in a temporary folder (e.g., using pytest fixtures) would cause unnecessary overhead.

Some of these are fixable upstream but since it is fairly simple to replicate the functionality we need, it makes sense to avoid the dependency on `sandbox.sh`. By doing that we'll also be forced to train our CCF muscles which is a good thing.

### PR DESCRIPTION
**Pull request by plietar**
We replace sandbox.sh with our own Python implementation, in `test/infra/cchost.py`.

The functional tests can be called with a `--start-cchost`, in which case the test infrastructure will launch a start and configure a new cchost process for the duration of the test session. This provides tighter and more powerful integration with the tests and opens up the door for new possibilities, such as starting an isolated cchost process for certain tests or performing a disaster recovery on the service. It is also friendlier to use, as a single command and console output manages everything.

It is still possible to run the tests against an external service, which is necessary to run the Docker based test jobs. While most test support this scenario, in the future individual tests may be annotated with `needs_cchost` and will be skipped if using an external service.

The `cchost.py` file may also be executed as a standalone script, which is useful for local development. The `run.sh` script has been updated to use this script instead of `sandbox.sh`.

Fixes #6.

### CODE REVIEW

**Comment by plietar** [comment]
This is good for review now. One remaining question is whether `--start-cchost` should be the default or not (we'd have a `--use-external-service` instead).

From my experience using it so far, I've found this to be much more convenient than having to keep a separate terminal open for the `start.sh`, so I'd be inclined to say yes and make it the default but I'm open to other opinions.

File: .pipelines/ccf.yml [hunk 1]
```diff
@@ -12,8 +12,11 @@ steps:
       git clone --single-branch -b ccf-${{ parameters.Version }} https://github.com/microsoft/CCF $(Pipeline.Workspace)/CCF
     displayName: Checkout CCF ${{ parameters.Version }}
 
-  # This commit is in the main branch of CCF. Once it gets backported to 2.x we can bump the CCF version and stop patching.
-  - script: curl -L https://github.com/microsoft/CCF/commit/3c7f7cbaf6b27cff0cae61c02bda4af7fb6fff92.diff | patch -p1
+  # These are commits from the main branch of CCF to fix ASAN builds, but which
```
Comment (letmaik) [inline .pipelines/ccf.yml:15]: Seems unrelated to this PR.

**Review by letmaik — CHANGES_REQUESTED** [review]

File: .pipelines/ccf.yml (hunk 1 above)
Comment (plietar) [reply .pipelines/ccf.yml:15]: It is related. The `cchost.py` uses CCF's json logger (whereas we previously used the text logger only), and [at some point in that code](https://github.com/microsoft/CCF/blob/d0990ec76e86a8555f16d3354116623ef41b7334/include/ccf/ds/logger.h#L142) the enclave offset gets converted from a double to a size_t, which is UB if the value does not fit in a size_t (this includes any negatives).

So whereas with the text logger we only get some garbage values printed, with the JSON logger trips ubsan and makes the process crash. The second patch I added fixes that bug: https://github.com/microsoft/CCF/pull/3972

I could ask the CCF team to backport this to 2.x, but given that running ASAN in apps isn't really supported on that branch anyway, it seems unlikely.

File: pyscitt/pyscitt/client.py [hunk 2]
```diff
@@ -183,6 +183,10 @@ def request(
             else:
                 kwargs["auth"] = self.member_http_sig
 
+            # Content-length is necessary for signing, even on GET requests.
+            if method == "GET":
```
[the remark this answers is not in the document]
Comment (plietar) [reply pyscitt/pyscitt/client.py:187]: The `/gov/recovery_share` endpoint needs it. This was part of a follow up PR that adds disaster recovery tests and it slipped through when I split the work. I'll remove it from this PR.

File: test/infra/fixtures.py [hunk 3]
```diff
@@ -10,29 +13,165 @@
...
+        "--start-cchost",
+        action="store_true",
+        help="Start a cchost process managed by the test framework",
+    )
+    parser.addoption(
+        "--cchost-binary",
+        default="/opt/ccf/bin/cchost",
+        help="Path to the cchost binary. Requires --start-cchost.",
+    )
+    parser.addoption(
```
[the remark this answers is not in the document]
Comment (plietar) [reply test/infra/fixtures.py:121]: "pick any free port" is not a straightforward thing to do. We might be able to get cchost to pass port 0 to the `bind` syscall to let the OS pick it, but then we have no way of knowing its value. We could bind port 0 in the python code, check the port number, close the socket and give that to cchost but it's a very racy operation and gets tricky if we start running multiple cchost processes.

I'll have a look at the CCF infra see if there's a trick I'm not aware of to do this elegantly.

File: test/test_cli.py [hunk 4]
```diff
@@ -482,10 +482,11 @@ def test_trailing_text(
         ):
             update_scitt_constitution("")
 
-    @pytest.mark.xfail(
-        reason="Test does not work with sandbox.sh, only cchost",
-        raises=pytest.fail.Exception,
-    )
+    # sandbox.sh uses a special consitution that accepts proposals without even
```
[the remark this answers is not in the document]
Comment (plietar) [reply test/test_cli.py:485]: Well, if using an external process we have no way of knowing how it got started and what it's constitution looks like. There's nothing stopping you from using `sandbox.sh` and running the tests against it.

On the other hand, we could also decide that the integration tests just don't support `sandbox.sh` anymore (either `start.sh` or the Docker images are still fine).

File: test/infra/cchost.py [hunk 5]
```diff
@@ -0,0 +1,414 @@
...
+            start_new_session=True,
+            stdin=subprocess.DEVNULL,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+            env={
+                **os.environ,
+                # This is needed for the attested fetch script to know where
+                # to contact the CCF service. The script is started by cchost,
+                # so it inherits any of cchost's environment variables.
+                "CCF_SERVICE_PORT": str(self.rpc_port),
```
[the remark this answers is not in the document]
Comment (plietar) [reply test/infra/cchost.py:193]: Even the port number isn't guaranteed to be the same. It's quite possible that cchost inside a docker image is listening on port 8000 but that is exposed as port 443 to the outside world.

File: test/test_cli.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (letmaik) [reply test/test_cli.py:485]: I think it's fine if we stop supporting `sandbox.sh`.

File: test/infra/fixtures.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (plietar) [reply test/infra/fixtures.py:121]: Oh wow so cchost has a mechanism for doing this, by binding to port 0 and then writing the port number to a file: https://github.com/microsoft/CCF/pull/1173

I'll update to using this instead.

File: test/infra/cchost.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (letmaik) [reply test/infra/cchost.py:193]: Ah you're right. Hm, ok, let's leave it for now until we find a better solution.

[the remark this answers is not in the document]
Comment (plietar) [reply test/infra/cchost.py:193]: Heh, so this is now broken by the random port assignment, since we don't have a way of setting an environment variable with the value ahead of starting the process. I need to figure out if there's a way to get this port number from within the application to pass it as a callback URL.

File: test/infra/fixtures.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (plietar) [reply test/infra/fixtures.py:118]: Done. The python code uses the file that cchost creates to determine the random port assignment.

The did:web resolver module uses the NodeInfo stored in the KV to construct a callback URL that gets passed to the external fetch scripts.

**Comment by plietar** [comment]
There seems to be some flakiness associated with the prefix tree tests. I don't think this PR introduced the bug but merely surfaced it by changing timings.

Still need to investigate and fix it before this can be merged, as it would make CI way too unreliable otherwise.

**Comment by plietar** [comment]
The aforementioned flaky test will fixed by #26. I've cherry-picked the commit here, but that PR should get merged first then I'll rebase this.

File: run_functional_tests.sh [hunk 6]
```diff
@@ -22,42 +15,35 @@ fi
...
-timeout=120
-while ! curl -s -f -k $CCF_URL/app/parameters > /dev/null; do
-    echo "Waiting for service to be ready..."
-    sleep 1
-    timeout=$((timeout - 1))
-    if [ $timeout -eq 0 ]; then
-        echo "Service failed to become ready, exiting"
-        exit 1
-    fi
-done
```
Comment (letmaik) [inline run_functional_tests.sh:44]: Isn't this still needed for the docker case?

Comment (letmaik) [reply run_functional_tests.sh:44]: It does, but that's running in the background, so there's a race here.

Comment (plietar) [reply run_functional_tests.sh:44]: Yes you're right. Fixed it now.

**Review by letmaik — APPROVED** [review]
